pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+                        target: "runtime::tendermint-lc",
+                        "Cannot finalize light block {:?} because Light Client is not yet initialized",
+                        &untrusted_block,
+                    );
+                    fail!(<Error<T>>::NotInitialized);
+                }
+            };
+
+            let now = T::TimeProvider::now();
+            let now = Time::from_unix_timestamp(now.as_secs().try_into().unwrap(), 0).unwrap();
+
+            // println!("@ now {:?}", now);
+
+            let verdict = verify_light_block(
+                verifier,
+                untrusted_block.clone(),
+                most_recent_trusted_block,
+                &options,
+                now,
+            );
+
+            match verdict {
+                tendermint_light_client_verifier::Verdict::Success => {
+                    // update storage
+                    insert_light_block::<T>(hash, untrusted_block);
+                    log::info!(target: "runtime::tendermint-lc", "Successfully verified light block {:#?}", &hash);
+                    Self::deposit_event(Event::ImportedLightBlock(who, hash));
+                    Ok(())
+                }
+                tendermint_light_client_verifier::Verdict::NotEnoughTrust(voting_power_tally) => {
+                    log::warn!(target: "runtime::tendermint-lc", "Not enough voting power to accept the light block {:#?}, vote tally  {}", &hash, &voting_power_tally);
+                    fail!(<Error<T>>::NotEnoughTrust)
+                }
+                tendermint_light_client_verifier::Verdict::Invalid(why) => {
+                    log::warn!(target: "runtime::tendermint-lc", "Rejecting invalid light block {:#?} becasue {}", &hash, &why);
+
+                    fail!(<Error<T>>::InvalidBlock)
+                }
+            }
+        }
+
+        /// Halt or resume all light client operations
+        ///
+        /// Can only be called by root
+        #[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+        pub fn set_halted(origin: OriginFor<T>, halted: bool) -> DispatchResult {
+            ensure_root(origin)?;
+            <IsHalted<T>>::put(halted);
+
+            if halted {
+                log::warn!(target: "runtime::tendermint-lc", "Halting light client operations");
+                Self::deposit_event(Event::LightClientHalted);
+            } else {
+                log::warn!(target: "runtime::tendermint-lc", "Resuming light client operations");
+                Self::deposit_event(Event::LightClientResumed);
+            }
+
+            Ok(())
+        }
+    }
+
+    fn verify_light_block(
+        verifier: ProdVerifier,
+        untrusted_block: LightBlockStorage,
+        trusted_block: LightBlockStorage,
+        options: &Options,
+        now: Time,
+    ) -> tendermint_light_client_verifier::Verdict {
+        let untrusted_block: LightBlock = untrusted_block
+            .clone()
```

Comment:
Why do we need to clone it? Its ownership is moved to the method.

Reply:
fixed